### ISSUE TITLE
[Request] Added using childrens to EquipmentContainerSetter

### ISSUE THREAD

**Issue by Eniotnacram** [suriyun-production/mmorpg-kit-docs#2204]
"**OBSOLETE**" New modifications **below** (Second post) ↓

I've expanded the EquipmentContainer script to make it able to use groups instead of a single gameobject.

**Why?**

This modification make it really easy to have modular characters in our projects, for example (Hand_R, Hand_L, Leg_L.. Etc.) It let the user put multiples gameobjects in a single index of an equip socket "Skin" in my case. This looks like this:

![Screenshot 2023-08-08 212740](https://github.com/suriyun-production/mmorpg-kit-docs/assets/46734935/f23e29a5-eb08-4a68-ba63-46e7fd896338)
So using this, I am able to use the  PlayerCharacterBodyPartComponent to trigger MULTIPLES meshes using only 1 index. For this example, index 0 is my Skin. The Skin socket have 9 gameobjects, when I want to change the color of my skin, it changes the color of all 9 gameobjects in index 0.

Ive also created an array for the defaultModel as it is also really helpfull for the same thing.

It works really well with the latest 1.84 release with the new PlayerCharacterBodyPartComponent.

Again, if possible, I would really like this to be implemented in the CORE, as I think it could be really usefull to anyone with the kit and only makes the kit more modulable!

There are all the scripts that have been modified for this: https://gist.github.com/Eniotnacram/936bb75e3a180be9a9d434b0bb991940

So again, it would be really nice to add this to the CORE kit. Im pretty sure I've modified every scripts related to this but I could be wrong.

Thank you again, really nice work on your latest release and I'm here if you need more information.

**Comment by insthync** [comment]
How about other customers' projects? I think it can break other customer projects, wouldn't it?

**Comment by Eniotnacram** [comment]
> How about other customers' projects? I think it can break other customer projects, wouldn't it?

No, It can be use exactly the same, as the original. If the array has only one index used, it act exactly the same as a if there is only 1 game object.

It could maybe remove the existing gameobjects from the script as we change it for a single gameobject to an array. If it would remove the existing gameobject of existing projects, we could add a flag to use groups (array) or not. And by default it's disabled. So it won't do anything to existing projects?

**Comment by Eniotnacram** [comment]
Yes its confirmed, when importing the new scripts to a fresh project it will remove the gameobjects referenced in the "Playable Character Model" > Equipment Containers. So to simply prevent this, we should add a check to enable or not the arrays. Exactly like the Custom Description task you added for the quests.

Again, I wouldnt mind keeping this for myself and creating an addon, but I think this upgrade is not user specific and could be helpfull to anyone using the kit, especially with the newest character customization impleted in 1.84.

**Comment by Eniotnacram** [comment]
Note that my current implementatio doesnt work currently if there is multiples arrays in the instantiated objects (currently trying to fix)

**Comment by Eniotnacram** [comment]
If I pay would you be able to add this properly @insthync? 

If we keep the default variables and we add new variables and a bool to enable this new functionnality it will not break other customer projects.

The core logic is already there in the scripts ive sent to you but I've not managed to add the bool flag and all.

Let me know, it would be very appreciated.

**Comment by Eniotnacram** [comment]
I found a really really simpler fix for this usecase, instead of all of this, i simply modifed the EquipmentContainerSetter to use the children of the gameobject refer. So I can have an empty gameobject called Skin. This gameobject has 7 children gameobject (renderer) for each body parts and I simply need to drag the Parent gameobject skin to the array to make it work.

So this method is really really simpler than the other and it doesnt impact any existing projects.

I will be sending the modified scripts here:

**Comment by insthync** [comment]
Okay

**Comment by LordPhrozen** [comment]
@insthync this sounds like a great thing to support out of the box with the kit, I'm sure we and other would use it.

**Comment by Eniotnacram** [comment]
Could potentially add to the "PlayerCharacterBodyPartComponent" that it only changes the "FIRST" material on the mesh. So IF multiple materials on a single mesh, it will not remove all of them and only add one. It will only modify the first material. (Usually the first material is the MAIN material of the mesh).

Exemple:
        private void OnShowEquipmentModel(EquipmentModel model, GameObject modelObject, BaseEquipmentEntity equipmentEntity, EquipmentContainer equipmentContainer)
        {
            // Get mesh's material to change color
            if (model == null || modelObject == null)
                return;

            if (model.indexOfModel < 0 || options[_currentModelIndex].colors.Length <= 0 || model.indexOfModel >= options[_currentModelIndex].colors[_currentColorIndex].ModelColorSettings.Length)
                return;

            // Fetch all Renderer components in the modelObject and its children
            Renderer[] renderers = modelObject.GetComponentsInChildren<Renderer>();
            foreach (Renderer renderer in renderers)
            {
                Material[] currentMaterials = renderer.materials;
                if (options[_currentModelIndex].colors[_currentColorIndex].ModelColorSettings[model.indexOfModel].materials.Length > 0)
                {
                    currentMaterials[0] = options[_currentModelIndex].colors[_currentColorIndex].ModelColorSettings[model.indexOfModel].materials[0];
                    renderer.materials = currentMaterials;
                }
            }
        }

**Comment by insthync** [comment]
How abount grand-grand-child?

**Comment by insthync** [comment]
I won't accept your changes, it will break other projects, you have to learn more about how `EquipmentModel` -> `instantiatedObjectIndex` is being used.

**Comment by insthync** [comment]
@Eniotnacram Please review latest commit "feat: multiple equipment model objects setting
"

**Comment by Eniotnacram** [comment]
Currently reviewing

**Comment by Eniotnacram** [comment]
Reviewed the latest commit, that looks like exactly what's needed. Thank you very much and sorry for my attempt to provide those scripts. I really dont want to break any other projects.

I love the way you implemented it.

**Comment by Eniotnacram** [comment]
@insthync 

Im not sure to understand the changes properly. I see the Instantiated Object Groups in the PlayableCharacterModel but there is no way to use them. For example in the PlayerCharacterBodypartComponent, there is no way to be able to use the "Instantiated Object Groups", no bool or nothing, so by default it will use the "normal" array of the Instantiated Objects. Am I missing something or it is not implemented?

**Comment by insthync** [comment]
Don't have to add condition if we can use length/size of areay/list to detect that we are going to use the array/list or not.
 
Do you just worried, or you found any issues?

**Comment by Eniotnacram** [comment]
I've tried to use it, with the current ArmorItem or the PlayerCharacterBodypartComponent, and there is no way to use them.

For example,  For my Skin Equip socket, The default model is empty and the instantiated objects list is also empty.

I have populated to Instantiated Object Groups with 7 gameobjects in "Element 0".

They are not used. I cannot change the material of those gameobjects using the PlayerCharacterBodypartComponent.
